### ISSUE TITLE
MOSEK and CBC Return Worse Solutions 

### ISSUE THREAD

**Issue by adishavit** [cvxgrp/cvxpy#1075]
I’m comparing solvers on a MILP problem with Gurobi, GLPK_MI, MOSEK and CBC. 
The first two return an optimal result with optimal status (the problem is always feasible). The objective maximizes a sum of integers. 
The latter two: MOSEK and CBC return a solution deemed `optimal` but obviously worse than the other two. These return a smaller integral value. 

Is there any way I can get a deeper understanding into why this happens (or how it can happen)?
Maybe it will lead me to tweaking my problem (or report a bug in these solvers). 

The problem has about 1000 constrains and variables.

**Comment by rileyjmurray** [comment]
It will be very hard for us to help without seeing the code used to generate the problem. The MOSEK and CBC interfaces are independent of each other, so it’s strange that the same bug would happen with both of them.

If you can’t share your data with us due to confidentiality reasons, and you have a commercial license for MOSEK, they may be able to provide the detailed support you need.

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by adishavit** [comment]
The code to generate the problem itself is quite complex and the problem is built up incrementally. It took quite a lot of trial and error, delta-debugging style, to trim the problem down but still have the problem manifested. 

If there is a way to somehow export the problem in some standard form, I could upload that I guess. 

To be clear CBC and MOSEK do not give the same result. They just both fail to provide a solution as good as the others.

**Comment by rileyjmurray** [comment]
I see. There are various points where a "standard form" can be exported. The simplest thing is to just pickle the entire cvxpy Problem object.

Is your problem using Parameter objects?

**Comment by adishavit** [comment]
No Parameters. 
Would `problem.get_problem_data()` help?
Do you have a pickle snippet (I’ve never used it before)?

**Comment by adishavit** [comment]
I can pickle the problem and upload it somewhere.

**Comment by rileyjmurray** [comment]
If you stick it on a public Google Drive or Dropbox folder that's fine. You could also try emailing it to me (you might have to move it to a zip file first) at ``rjmurray@berkeley.edu``.

**Comment by aszekMosek** [comment]
If you use ``save_file="dump.mps"`` with the MOSEK solver (https://docs.mosek.com/9.2/faq/faq.html#cvxpy) then you will get the problem in MPS format. Hard to get more standard than that. Then you can solve it with various solvers. Of course I assume that there is no bug in CVXPY and that the MPS file will in fact represent your problem. I would suspect the more likely scenario that there is something numerically fishy in your data that makes other solvers go astray, for instance by solving the relaxations poorly. You are also welcome to send the MPS file to ``support@mosek.com`` (together with the "more optimal" solution, if possible).

**Comment by aszekMosek** [comment]
The MPS file dumped from MOSEK I received from @adishavit gives the same answer in MOSEK and in another solver, it has no obvious numerical issues and solves easily. So my previous guess was wrong. Something else is going on.

**Comment by rileyjmurray** [comment]
@adishavit, I ran the problem you sent over. Strangely, when I tried to re-solve the unpickled problems, I ran into attribute errors about ``Problem._cache``. Based on CVXPY source code I was able to inject a fix so that I could still run the problem. With that fix, MOSEK, GUROBI, and GLPK all returned the same solution. Here is the code I ran:
```
import pickle
import cvxpy as cp

loc = '/Users/RJMurray/Library/Preferences/PyCharm2019.2/scratches/problem.pickle'
file = open(loc, 'rb')
prob = pickle.load(file)
file.close()

try:
    prob.solve(solver='MOSEK', verbose=True)
except AttributeError as err:
    separator_string = '\n' + ('-' * 80) + '\n'
    print(separator_string)
    print(err)
    print(separator_string)
    if '_cache' in str(err):
        prob.__dict__['_cache'] = cp.problems.problem.Cache()
    prob.solve(solver='MOSEK', verbose=True)
    print(separator_string)
    prob.solve(solver='GUROBI', verbose=True)
    print(separator_string)
    prob.solve(solver='GLPK_MI', verbose=True)
    print(separator_string)
```
Here is the output from each call to ``prob.solve``:
```
--------------------------------------------------------------------------------

'Problem' object has no attribute '_cache'

--------------------------------------------------------------------------------
Problem
  Name                   :                 
  Objective sense        : min             
  Type                   : LO (linear optimization problem)
  Constraints            : 1254            
  Cones                  : 0               
  Scalar variables       : 917             
  Matrix variables       : 0               
  Integer variables      : 917             
Optimizer started.
Mixed integer optimizer started.
Threads used: 4
Presolve started.
Presolve terminated. Time = 0.08
Presolved problem: 913 variables, 1243 constraints, 13797 non-zeros
Presolved problem: 0 general integer, 913 binary, 0 continuous
Clique table size: 4566
BRANCHES RELAXS   ACT_NDS  DEPTH    BEST_INT_OBJ         BEST_RELAX_OBJ       REL_GAP(%)  TIME  
0        0        1        0        0.0000000000e+00     NA                   NA          0.1   
0        1        1        0        0.0000000000e+00     -3.5000000000e+01    3.50e+13    0.1   
0        1        1        0        -3.5000000000e+01    -3.5000000000e+01    0.00e+00    0.2   
An optimal solution satisfying the relative gap tolerance of 1.00e-02(%) has been located.
The relative gap is 0.00e+00(%).
An optimal solution satisfying the absolute gap tolerance of 0.00e+00 has been located.
The absolute gap is 0.00e+00.
Objective of best integer solution : -3.500000000000e+01     
Best objective bound               : -3.500000000000e+01     
Construct solution objective       : Not employed
User objective cut value           : Not employed
Number of cuts generated           : 0
Number of branches                 : 0
Number of relaxations solved       : 1
Number of interior point iterations: 11
Number of simplex iterations       : 0
Time spend presolving the root     : 0.08
Time spend optimizing the root     : 0.06
Mixed integer optimizer terminated. Time: 0.16
Optimizer terminated. Time: 0.17    
Integer solution solution summary
  Problem status  : PRIMAL_FEASIBLE
  Solution status : INTEGER_OPTIMAL
  Primal.  obj: -3.5000000000e+01   nrm: 2e+01    Viol.  con: 2e-15    var: 3e-16    itg: 1e-15  

--------------------------------------------------------------------------------

Using license file /Users/RJMurray/gurobi.lic
Academic license - for non-commercial use only
Parameter OutputFlag unchanged
   Value: 1  Min: 0  Max: 1  Default: 1
Changed value of parameter QCPDual to 1
   Prev: 0  Min: 0  Max: 1  Default: 0
Gurobi Optimizer version 9.0.1 build v9.0.1rc0 (mac64)
Optimize a model with 1254 rows, 917 columns and 14168 nonzeros
Model fingerprint: 0x5f29fb71
Variable types: 0 continuous, 917 integer (917 binary)
Coefficient statistics:
  Matrix range     [1e+00, 2e+01]
  Objective range  [1e+00, 1e+00]
  Bounds range     [1e+00, 1e+00]
  RHS range        [1e+00, 2e+02]
Found heuristic solution: objective 0.0000000
Presolve removed 566 rows and 421 columns
Presolve time: 0.08s
Presolved: 688 rows, 496 columns, 8875 nonzeros
Found heuristic solution: objective -29.0000000
Variable types: 0 continuous, 496 integer (496 binary)
Root relaxation: objective -3.500000e+01, 295 iterations, 0.01 seconds
    Nodes    |    Current Node    |     Objective Bounds      |     Work
 Expl Unexpl |  Obj  Depth IntInf | Incumbent    BestBd   Gap | It/Node Time
H    0     0                     -35.0000000 -221.00000   531%     -    0s
     0     0          -    0       -35.00000  -35.00000  0.00%     -    0s
Explored 0 nodes (517 simplex iterations) in 0.10 seconds
Thread count was 4 (of 4 available processors)
Solution count 3: -35 -29 0 
No other solutions better than -35
Optimal solution found (tolerance 1.00e-04)
Best objective -3.500000000000e+01, best bound -3.500000000000e+01, gap 0.0000%

--------------------------------------------------------------------------------

*     0: obj =   0.000000000e+00 inf =   0.000e+00 (327)
*   709: obj =  -3.500000000e+01 inf =   1.070e-13 (0) 3
Long-step dual simplex will be used
+   709: mip =     not found yet >=              -inf        (1; 0)
+  1246: >>>>>  -3.500000000e+01 >=  -3.500000000e+01   0.0% (13; 1)
+  1246: mip =  -3.500000000e+01 >=     tree is empty   0.0% (0; 27)

--------------------------------------------------------------------------------
```
I also inspected the objective function. This takes a bit of digging into cvxpy objects, but it seems the objective function is actually to maximize a sum of 327 binary varibales. So, maybe the "fix" I employed for the cache somehow substantially changed the problem data? That would surprise me, but it's not outside the realm of possibility.

**Comment by adishavit** [comment]
@rileyjmurray @aszekMosek thank you so much for your prompt and efficient support!

After getting your results I went back to my code to investigate how I read back the solution results.
For the integer variables, I changed an `int()` cast to `round()`.  
I also found a bug where I used `int(res.value)` instead of `round(res.value[0])` (where `res` is a `cp.Variable`).
This change made all solvers (including CBC) agree and return 35 as the optimal value.  

I am still unsure why different solvers disagreed or why two actually yielded the correct result.  
Maybe this bug/fix can help shed light on some internal API issue.
